File: app/src/main/java/com/msd/notebook/fragments/InstructorHomeFragment.java
package com.msd.notebook.fragments;

import static android.app.Activity.RESULT_OK;
import static com.msd.notebook.common.Constants.DOCUMENT_REQUEST;
import static com.msd.notebook.common.Constants.FILE_EXT;
import static com.msd.notebook.common.Constants.FILE_NAME;
import static com.msd.notebook.common.Constants.FILE_URL;
import static com.msd.notebook.common.Constants.FIRESTORE_DOC_ID;
import static com.msd.notebook.common.Constants.INSTRUCTOR;
import static com.msd.notebook.common.Constants.YOUR_INSTRUCTORS;
import static com.msd.notebook.common.Constants.YOUR_UPLOADS;

import android.app.ProgressDialog;
import android.content.ContentResolver;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.MimeTypeMap;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.FirebaseApp;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.OnProgressListener;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;
import com.msd.notebook.adapter.FileAdapter;
import com.msd.notebook.common.PreferenceClass;
import com.msd.notebook.common.ProgressBarClass;
import com.msd.notebook.databinding.FragmentHomeBinding;
import com.msd.notebook.models.InstructorFiles;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class InstructorHomeFragment extends Fragment {

    FragmentHomeBinding binding;
    PreferenceClass preferenceClass;
    FirebaseFirestore db = FirebaseFirestore.getInstance();
    List<InstructorFiles> filesList = new ArrayList<>();


    FileAdapter adapter;

    public InstructorHomeFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        FirebaseApp.initializeApp(getContext());

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentHomeBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        preferenceClass = new PreferenceClass(getContext());

        binding.addFileFloat.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                openingDocumentintent();
            }
        });

        adapter = new FileAdapter(getContext(), true, new FileAdapter.FileBtnClick() {
            @Override
            public void btnClick(InstructorFiles file) {
                //delete code
                removeProductFromFirestore(file);
            }
        });

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext());
        binding.filesRecyclerView.setLayoutManager(linearLayoutManager);
        binding.filesRecyclerView.setAdapter(adapter);

        getYourFiles();

    }

    private void getYourFiles() {

        String userDoc = preferenceClass.getString(FIRESTORE_DOC_ID);

        db.collection(INSTRUCTOR)
                .document(userDoc)
                .collection(YOUR_UPLOADS)
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        ProgressBarClass.getInstance().dismissProgress();
                        if (task.isSuccessful()) {
                            for (QueryDocumentSnapshot document : task.getResult()) {
                                InstructorFiles file = new InstructorFiles(document.getId(),
                                        document.getData().get(FILE_NAME).toString(),
                                        document.getData().get(FILE_URL).toString(),
                                        document.getData().get(FILE_EXT).toString()
                                );
                                filesList.add(file);
                            }
                            adapter.setFiles(filesList);
                        } else {
                            Log.e("HomeFragment", "Error getting documents.", task.getException());
                        }

//                        if (filesList.isEmpty()) {
//                            binding.fileText.setText("Your list is empty");
//                        }
                    }
                });
    }

    private void openingDocumentintent() {
        Intent chooseFile = new Intent(Intent.ACTION_GET_CONTENT);
        chooseFile.setType("*/*");
        chooseFile = Intent.createChooser(chooseFile, "Choose a file");
        startActivityForResult(chooseFile, DOCUMENT_REQUEST);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == DOCUMENT_REQUEST && resultCode == RESULT_OK) {
            uploadFileToFirebase(data);
        }
    }

    private void uploadFileToFirebase(Intent data) {
        ProgressDialog progressDialog = new ProgressDialog(getContext());
        progressDialog.setTitle("Uploading...");
        progressDialog.show();
        FirebaseStorage storage = FirebaseStorage.getInstance();
        StorageReference storageRef = storage.getReference();

        Uri uri = data.getData();
        File file = new File(uri.getPath());

        String instructorDocId = preferenceClass.getString(FIRESTORE_DOC_ID);

        String fileName = System.currentTimeMillis() + "";

        StorageReference reference = storageRef.child(instructorDocId + "/" +
                fileName + "." + getfileExtension(uri));
        reference.putFile(data.getData())
                .addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
                    @Override
                    public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {
                        Task<Uri> uriTask = taskSnapshot.getStorage().getDownloadUrl();
                        while (!uriTask.isComplete()) ;

                        Uri url = uriTask.getResult();

                        saveFileInFireStoreUploads(fileName, url, getfileExtension(uri));

                        Toast.makeText(getContext(), "File Uploaded", Toast.LENGTH_SHORT).show();
                        progressDialog.dismiss();
                    }
                }).addOnProgressListener(new OnProgressListener<UploadTask.TaskSnapshot>() {
                    @Override
                    public void onProgress(@NonNull UploadTask.TaskSnapshot snapshot) {
                        double progress = (100.0 * snapshot.getBytesTransferred()) / snapshot.getTotalByteCount();
                        progressDialog.setMessage("Uploaded: " + (int) progress + "%");
                    }
                });
    }

    private void removeProductFromFirestore(InstructorFiles file) {
        String userDoc = preferenceClass.getString(FIRESTORE_DOC_ID);

        db.collection(INSTRUCTOR)
                .document(userDoc)
                .collection(YOUR_UPLOADS)
                .document(file.getId())
                .delete()
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void unused) {
                        Toast.makeText(getContext(),
                                "File Removed ", Toast.LENGTH_SHORT).show();
                        filesList.clear();
                        getYourFiles();
                    }
                }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Toast.makeText(getContext(),
                                "Error, Please try again", Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void saveFileInFireStoreUploads(String fileName, Uri url, String ext) {
        Map<String, Object> file = new HashMap<>();
        file.put(FILE_NAME, fileName);
        file.put(FILE_URL, url);
        file.put(FILE_EXT, ext);

        String userDoc = preferenceClass.getString(FIRESTORE_DOC_ID);

        db.collection(INSTRUCTOR)
                .document(userDoc)
                .collection(YOUR_UPLOADS)
                .add(file)
                .addOnSuccessListener(new OnSuccessListener<DocumentReference>() {
                    @Override
                    public void onSuccess(DocumentReference documentReference) {
                        Toast.makeText(getContext(), "File Added", Toast.LENGTH_SHORT).show();

                        filesList.clear();
                        getYourFiles();
                    }
                }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Toast.makeText(getContext(), "Error, Please try again", Toast.LENGTH_SHORT).show();
                    }
                });
    }

    //https://stackoverflow.com/questions/37951869/how-to-get-the-file-extension-in-android
    private String getfileExtension(Uri uri) {
        String extension;
        ContentResolver contentResolver = getContext().getContentResolver();
        MimeTypeMap mimeTypeMap = MimeTypeMap.getSingleton();
        extension = mimeTypeMap.getExtensionFromMimeType(contentResolver.getType(uri));
        return extension;
    }
}

File: app/src/main/java/com/msd/notebook/view/fragments/InstructorHomeFragment.kt
package com.msd.notebook.view.fragments

import android.app.Activity
import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.OpenableColumns
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.MimeTypeMap
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.gms.tasks.OnSuccessListener
import com.google.firebase.FirebaseApp
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.storage.FirebaseStorage
import com.msd.notebook.common.Constants
import com.msd.notebook.common.PreferenceClass
import com.msd.notebook.common.ProgressBarClass
import com.msd.notebook.databinding.FragmentHomeBinding
import com.msd.notebook.models.InstructorFiles
import com.msd.notebook.view.adapter.FileAdapter
import com.msd.notebook.view.adapter.FileAdapter.FileBtnClick
import java.io.File

class InstructorHomeFragment : Fragment() {
    var binding: FragmentHomeBinding? = null
    var preferenceClass: PreferenceClass? = null
    var db = FirebaseFirestore.getInstance()
    var filesList: ArrayList<InstructorFiles> = ArrayList()
    var adapter: FileAdapter? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        FirebaseApp.initializeApp(requireContext())
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        binding = FragmentHomeBinding.inflate(inflater, container, false)
        return binding!!.getRoot()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        preferenceClass = PreferenceClass(context)
        binding!!.addFileFloat.setOnClickListener { openingDocumentintent() }
        adapter = FileAdapter(requireContext()!!, true, object : FileBtnClick {
            override fun btnClick(file: InstructorFiles?) {
                //delete code
                removeProductFromFirestore(file)
            }

            override fun cardClcik(file: InstructorFiles?) {
                TODO("Not yet implemented")
            }
        })
        val linearLayoutManager = LinearLayoutManager(context)
        binding!!.filesRecyclerView.setLayoutManager(linearLayoutManager)
        binding!!.filesRecyclerView.setAdapter(adapter)
        yourFiles()
    }

    fun yourFiles() {
        val userDoc = preferenceClass!!.getString(Constants.FIRESTORE_DOC_ID)
        db.collection(Constants.INSTRUCTOR)
            .document(userDoc!!)
            .collection(Constants.YOUR_UPLOADS)
            .get()
            .addOnCompleteListener { task ->
                ProgressBarClass.Companion.instance.dismissProgress()
                if (task.isSuccessful) {
                    for (document in task.result) {
                        val file = InstructorFiles(
                            document.id,
                            document.getData()[Constants.FILE_NAME].toString(),
                            document.getData()[Constants.FILE_URL].toString(),
                            document.getData()[Constants.FILE_EXT].toString()
                        )
                        filesList.add(file)
                    }
                    adapter!!.updateFiles(filesList)
                } else {
                    Log.e("HomeFragment", "Error getting documents.", task.exception)
                }

//                        if (filesList.isEmpty()) {
//                            binding.fileText.setText("Your list is empty");
//                        }
            }
    }

    private fun openingDocumentintent() {
        var chooseFile = Intent(Intent.ACTION_GET_CONTENT)
        chooseFile.setType("*/*")
        chooseFile = Intent.createChooser(chooseFile, "Choose a file")
        startActivityForResult(chooseFile, Constants.DOCUMENT_REQUEST)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == Constants.DOCUMENT_REQUEST && resultCode == Activity.RESULT_OK) {
            uploadFileToFirebase(data)
        }
    }

    fun getFileName(context: Context?, uri: Uri): String? {
        var fileName: String? = ""
        if (uri.scheme == "content") {
            val cursor = context?.contentResolver?.query(uri, null, null, null, null)
            cursor?.use {
                if (it.moveToFirst()) {
                    val nameIndex = it.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                    fileName = it.getString(nameIndex)
                }
            }
        }
        if (fileName == null) {
            // Fallback: extract from the file path if it's not a content Uri
            fileName = uri.path?.substring(uri.path!!.lastIndexOf('/') + 1)
        }
        return fileName
    }

    private fun uploadFileToFirebase(data: Intent?) {
        val progressDialog = ProgressDialog(context)
        progressDialog.setTitle("Uploading...")
        progressDialog.show()
        val storage = FirebaseStorage.getInstance()
        val storageRef = storage.getReference()
        val uri = data!!.data
        val file = File(uri!!.path)
        val instructorDocId = preferenceClass!!.getString(Constants.FIRESTORE_DOC_ID)
        val fileName = getFileName(context, uri)
        val reference = storageRef.child(
            instructorDocId + "/" +
                    fileName + "." + getfileExtension(uri)
        )
        reference.putFile(data.data!!)
            .addOnSuccessListener { taskSnapshot ->
                val uriTask = taskSnapshot.storage.getDownloadUrl()
                while (!uriTask.isComplete);
                val url = uriTask.result
                if (fileName != null) {
                    saveFileInFireStoreUploads(fileName, url, getfileExtension(uri))
                }
                Toast.makeText(context, "File Uploaded", Toast.LENGTH_SHORT).show()
                progressDialog.dismiss()
            }
            .addOnProgressListener { snapshot ->
                val progress = 100.0 * snapshot.bytesTransferred / snapshot.totalByteCount
                progressDialog.setMessage("Uploaded: " + progress.toInt() + "%")
            }
    }

    private fun removeProductFromFirestore(file: InstructorFiles?) {
        val userDoc = preferenceClass!!.getString(Constants.FIRESTORE_DOC_ID)
        db.collection(Constants.INSTRUCTOR)
            .document(userDoc!!)
            .collection(Constants.YOUR_UPLOADS)
            .document(file!!.id)
            .delete()
            .addOnSuccessListener(object : OnSuccessListener<Void?> {
                override fun onSuccess(unused: Void?) {
                    Toast.makeText(
                        context,
                        "File Removed ", Toast.LENGTH_SHORT
                    ).show()
                    filesList.clear()
                    yourFiles()
                }
            }).addOnFailureListener {
                Toast.makeText(
                    context,
                    "Error, Please try again", Toast.LENGTH_SHORT
                ).show()
            }
    }

    private fun saveFileInFireStoreUploads(fileName: String, url: Uri, ext: String?) {
        val file: MutableMap<String?, Any?> = HashMap()
        file[Constants.FILE_NAME] = fileName
        file[Constants.FILE_URL] = url
        file[Constants.FILE_EXT] = ext
        val userDoc = preferenceClass!!.getString(Constants.FIRESTORE_DOC_ID)
        db.collection(Constants.INSTRUCTOR)
            .document(userDoc!!)
            .collection(Constants.YOUR_UPLOADS)
            .add(file)
            .addOnSuccessListener(object : OnSuccessListener<DocumentReference?> {
                override fun onSuccess(documentReference: DocumentReference?) {
                    Toast.makeText(context, "File Added", Toast.LENGTH_SHORT).show()
                    filesList.clear()
                    yourFiles()
                }
            }).addOnFailureListener {
                Toast.makeText(
                    context,
                    "Error, Please try again",
                    Toast.LENGTH_SHORT
                ).show()
            }
    }

    //https://stackoverflow.com/questions/37951869/how-to-get-the-file-extension-in-android
    private fun getfileExtension(uri: Uri?): String? {
        val extension: String
        val contentResolver = requireContext()!!.contentResolver
        val mimeTypeMap = MimeTypeMap.getSingleton()
        extension = mimeTypeMap.getExtensionFromMimeType(contentResolver.getType(uri!!)).toString()
        return extension
    }
}